File: mobile-apps/alloy-android/app/src/main/kotlin/com/sharca/alloy/homeassistant/HomeAssistantConnection.kt
package com.sharca.alloy.homeassistant

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * Home Assistant REST + WebSocket client. One instance per HA
 * connection (host + token). Used by both [HomeAssistantTarget] (for
 * command + push) and [HomeAssistantDiscovery] (for entity
 * enumeration).
 *
 * See `mobile-apps/specs/home-assistant-api.md`.
 */
class HomeAssistantConnection(val config: Config) {

    @Serializable
    data class Config(val baseURL: String, val token: String)

    private val client = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun ping() {
        val request = Request.Builder()
            .url(config.baseURL.trimEnd('/') + "/api/")
            .addHeader("Authorization", "Bearer ${config.token}")
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw HomeAssistantException.AuthFailed
        }
    }

    suspend fun fetchStates(): List<HAState> {
        val request = Request.Builder()
            .url(config.baseURL.trimEnd('/') + "/api/states")
            .addHeader("Authorization", "Bearer ${config.token}")
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw HomeAssistantException.RequestFailed
            val body = response.body?.string() ?: return emptyList()
            val elements = json.parseToJsonElement(body).let { it as? kotlinx.serialization.json.JsonArray }
                ?: return emptyList()
            return elements.mapNotNull { parseState(it) }
        }
    }

    suspend fun setVolumeLevel(entityId: String, level: Double) {
        val body = """
            {"entity_id":"$entityId","volume_level":$level}
        """.trimIndent().toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url(config.baseURL.trimEnd('/') + "/api/services/media_player/volume_set")
            .addHeader("Authorization", "Bearer ${config.token}")
            .post(body)
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw HomeAssistantException.RequestFailed
        }
    }

    val websocketURL: String
        get() {
            val base = config.baseURL.trimEnd('/')
            return when {
                base.startsWith("https://") -> "wss://" + base.removePrefix("https://") + "/api/websocket"
                base.startsWith("http://")  -> "ws://"  + base.removePrefix("http://")  + "/api/websocket"
                else -> "ws://$base/api/websocket"
            }
        }

    private fun parseState(element: JsonElement): HAState? {
        val obj = element.jsonObject
        val entityId = obj["entity_id"]?.jsonPrimitive?.contentOrNull ?: return null
        val state = obj["state"]?.jsonPrimitive?.contentOrNull ?: ""
        val attributes = obj["attributes"]?.jsonObject
        val friendly = attributes?.get("friendly_name")?.jsonPrimitive?.contentOrNull
        val deviceClass = attributes?.get("device_class")?.jsonPrimitive?.contentOrNull
        val volume = attributes?.get("volume_level")?.jsonPrimitive?.doubleOrNull
        return HAState(entityId, state, friendly, deviceClass, volume)
    }

    val baseHost: String get() = config.baseURL
        .removePrefix("https://").removePrefix("http://").substringBefore(':').substringBefore('/')

    val basePort: Int get() = runCatching {
        java.net.URL(config.baseURL).port.takeIf { it != -1 } ?: 8123
    }.getOrDefault(8123)
}

data class HAState(
    val entityId: String,
    val state: String,
    val friendlyName: String?,
    val deviceClass: String?,
    val volumeLevel: Double?
)

sealed class HomeAssistantException(message: String) : Exception(message) {
    data object AuthFailed : HomeAssistantException("Home Assistant rejected the token.")
    data object RequestFailed : HomeAssistantException("Home Assistant request failed.")
    data object WebSocketFailed : HomeAssistantException("Lost connection to Home Assistant.")
}

File: mobile-apps/alloy-ios/Alloy/HomeAssistant/HomeAssistantConnection.swift
import Foundation

/// Home Assistant REST + WebSocket client. One instance per HA
/// connection (host + token). Used by both `HomeAssistantTarget`
/// (for command + push) and `HomeAssistantDiscovery` (for entity
/// enumeration).
///
/// See `mobile-apps/specs/home-assistant-api.md`.
actor HomeAssistantConnection {

    struct Config: Equatable, Sendable, Codable {
        var baseURL: URL   // e.g. https://ha.local:8123
        var token: String
    }

    private let config: Config
    private let session: URLSession

    init(config: Config, session: URLSession = .shared) {
        self.config = config
        self.session = session
    }

    // MARK: - REST

    /// `GET /api/` — trivial token test used by the HA connect screen.
    func ping() async throws {
        var req = URLRequest(url: config.baseURL.appendingPathComponent("api/"))
        req.setValue("Bearer \(config.token)", forHTTPHeaderField: "Authorization")
        let (_, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw HomeAssistantError.authFailed
        }
    }

    /// `GET /api/states` — all entity states.
    func fetchStates() async throws -> [HAState] {
        var req = URLRequest(url: config.baseURL.appendingPathComponent("api/states"))
        req.setValue("Bearer \(config.token)", forHTTPHeaderField: "Authorization")
        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw HomeAssistantError.requestFailed
        }
        return try JSONDecoder().decode([HAState].self, from: data)
    }

    /// `POST /api/services/media_player/volume_set` with a 0.0–1.0 level.
    func setVolumeLevel(entityId: String, level: Double) async throws {
        let url = config.baseURL.appendingPathComponent("api/services/media_player/volume_set")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("Bearer \(config.token)", forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body: [String: Any] = [
            "entity_id": entityId,
            "volume_level": level
        ]
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (_, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw HomeAssistantError.requestFailed
        }
    }

    var websocketURL: URL {
        var comps = URLComponents(url: config.baseURL, resolvingAgainstBaseURL: false) ?? URLComponents()
        comps.scheme = (comps.scheme == "https") ? "wss" : "ws"
        comps.path = "/api/websocket"
        return comps.url ?? config.baseURL.appendingPathComponent("api/websocket")
    }

    var token: String { config.token }
    var baseURL: URL { config.baseURL }
}

// MARK: - Models

struct HAState: Decodable, Sendable {
    let entity_id: String
    let state: String
    let attributes: [String: HAJSONValue]?

    var friendlyName: String? {
        if case let .string(s)? = attributes?["friendly_name"] { return s }
        return nil
    }

    var volumeLevel: Double? {
        if case let .number(d)? = attributes?["volume_level"] { return d }
        return nil
    }

    var deviceClass: String? {
        if case let .string(s)? = attributes?["device_class"] { return s }
        return nil
    }
}

/// Minimal JSON value type — we only decode the fields we care about.
enum HAJSONValue: Decodable, Sendable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case null
    case other

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() { self = .null; return }
        if let s = try? c.decode(String.self) { self = .string(s); return }
        if let d = try? c.decode(Double.self) { self = .number(d); return }
        if let b = try? c.decode(Bool.self) { self = .bool(b); return }
        self = .other
    }
}

enum HomeAssistantError: Error, LocalizedError {
    case authFailed
    case requestFailed
    case websocketFailed

    var errorDescription: String? {
        switch self {
        case .authFailed:      return "Home Assistant rejected the token."
        case .requestFailed:   return "Home Assistant request failed."
        case .websocketFailed: return "Lost connection to Home Assistant."
        }
    }
}
